src/main/java/org/folio/inventory/instanceingress/handler/CreateInstanceIngressEventHandler.java

```diff
@@ -0,0 +1,246 @@
+                                           IdStorageService idStorageService,
+                                           HttpClient httpClient,
+                                           Context context,
+                                           Storage storage) {
+    super(storage, precedingSucceedingTitlesHelper, mappingMetadataCache, idStorageService, null, httpClient);
+    this.context = context;
+    this.instanceCollection = storage.getInstanceCollection(context);
+  }
+
+  @Override
+  public CompletableFuture<Instance> handle(InstanceIngressEvent event) {
+    try {
+      LOGGER.info("Processing InstanceIngressEvent with id '{}' for instance creation", event.getId());
+      var future = new CompletableFuture<Instance>();
+      if (eventContainsNoData(event)) {
+        var message = format("InstanceIngressEvent message does not contain required data to create Instance for eventId: '%s'", event.getId());
+        LOGGER.error(message);
+        return CompletableFuture.failedFuture(new EventProcessingException(message));
+      }
+
+      var targetRecord = constructMarcBibRecord(event.getEventPayload());
+      var instanceId = ofNullable(event.getEventPayload().getSourceRecordIdentifier()).orElseGet(() -> getInstanceId(targetRecord));
+      idStorageService.store(targetRecord.getId(), instanceId, context.getTenantId())
+        .compose(res -> super.getMappingMetadataCache().getByRecordType(InstanceIngressEventConsumer.class.getSimpleName(), context, MARC_BIB_RECORD_TYPE))
+        .compose(metadataOptional -> metadataOptional.map(metadata -> prepareAndExecuteMapping(metadata, targetRecord, event, instanceId))
+          .orElseGet(() -> Future.failedFuture("MappingMetadata was not found for marc-bib record type")))
+        .compose(instance -> validateInstance(instance, event))
+        .compose(instance -> saveInstance(instance, event))
+        .onFailure(e -> {
+          if (!(e instanceof DuplicateEventException)) {
+            LOGGER.error(FAILURE, event.getId(), e);
+          }
+          future.completeExceptionally(e);
+        })
+        .onComplete(ar -> future.complete(ar.result()));
+      return future;
+    } catch (Exception e) {
+      LOGGER.error(FAILURE, event.getId(), e);
+      return CompletableFuture.failedFuture(e);
+    }
+  }
+
+  private boolean eventContainsNoData(InstanceIngressEvent event) {
+    return isNull(event.getEventPayload())
+      || isNull(event.getEventPayload().getSourceRecordObject())
+      || isNull(event.getEventPayload().getSourceType());
+  }
+
+  private Future<org.folio.Instance> prepareAndExecuteMapping(MappingMetadataDto mappingMetadata,
+                                                              Record targetRecord,
+                                                              InstanceIngressEvent event,
+                                                              String instanceId) {
+    return postSnapshotInSrsAndHandleResponse(targetRecord.getId())
```

Comment:
Maybe it is better to create snapshot right before saving the record in SRS? This operation logically belongs there and has no relation to the mapping of Instance

Reply:
That's correct, moved it to be executed just before saving to SRS